### DISCUSSION TITLE
Bytearray Question

### DISCUSSION THREAD

**Discussion by A622266** [micropython/micropython#10032] [Hardware & Peripherals] [answered]
I need to boot a peripheral device from an ``esp32s3``.  The peripheral device has a contiguous program memory organized as 40-bit instructions.  I have these as a comma delimited file from another tool as in:
```
0x00 , 0x00 , 0x00 , 0x00 , 0x00 , 

0xFE , 0xE0 , 0x00 , 0x00 , 0x00 , 

etc.
```
Is it correct for me to simply declare this as a ```bytearray```, and then write it to the peripheral over ``I2C``?
So I can just write:
```
program = bytearray([
0x00 , 0x00 , 0x00 , 0x00 , 0x00 , 

0xFE , 0xE0 , 0x00 , 0x00 , 0x00 , 

etc.
])
```
When I print ``program``, I don't see exactly the same values as the comma delimited file.  I think this is how bytes are printed, and is not unexpected, correct?  For example, ``0x20`` is printed as a blank.  Can I be confident that the ``bytearray`` values match those of the file, meaning I don't have the wrong idea to declare as a ``bytearray``as shown?  Or is there a way besides printing to verify that the program I am loading matches the one from comma delimited file?

**Comment by jimmo** [comment, marked as answer]
@A622266 Yep that's right, this is standard Python behavior -- bytes and bytearrays are printed as strings, where only the "unprintables" are escaped.

I would recommend instead making your tool print out the same format as Python prints

i.e.

`program = b'\x00\x00\x00\x00\x00\xfe\xe0\x00....'`

This has the added benefit of using less memory when loading your program, as the initial consutrction of the list uses 4x as much RAM (because the bytes will be promoted to integers, only to be truncated back down to bytes when creating the bytearray).

If you want to verify the data use

`print(program.hex(','))`
